---
title: "Ito_ImmunoSEQ"
author: "Mark Long"
date: "8/27/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r set wd}

# set global options so that graphics devices can be exported
options(bitmapType='cairo')

# set a mount variable
if(grepl("/home/mark", getwd())){
  mount <- "/run/user/1003/gvfs/sftp:host=u2,user=mlong"
} else {
  mount <- ""
}

# Set wd variable
wd <- "/mnt/lustre/users/mlong/Projects/Ito_ISIM/TCRseq"
   
```


```{r load libraries, message=FALSE}
require(ggplot2)
require(dplyr)
require(broom)
require(forcats)
require(reshape2)
require(pheatmap)
require(RColorBrewer)
require(vegan)
require(readxl)
```

  
  
# **TCR-seq** #
  
# **Method Overview** #

* TCR biology in adaptive immunity

Developing lymphocytes (progenitor) undergo a unique form of genetic recombination in the early stages of T and B cell maturation. This process is called V(D)J recombination, and occurs in the primary lymphoid organs (bone marrow - B cells; thymus - T cells). V(D)J recombination involves a random rearrangement of three separate gene segments; variable (V), joining (J) and diversity (D). The process ultimately results in a final nucleic acid sequence containing one of each V, D, and J segments. The unique VDJ gene produced through this process generates the variable region of both the beta (or heavy in B cells) and alpha (light chain in B cells, contain only V/J segments) chain polypeptides that in combination form a unique receptor protein, called a T or B cell receptor (TCR/BCR). Within the variable region of a TCR reside three complementarity-determining regions (CDRs), which ultimately is where the TCR interacts with its specific antigen. These regions are called the hypervariable regions, as a majority of the sequence variation is found here. The CDR1 and CDR2 are found within the V segment, while the CDR3 includes some of V, and all of D and J regions. CDR3 is the most variable and interacts with the target oligo-peptide lodged in the antigen-binding groove of the HLA molecule of an antigen-presenting cell. Since each TCR is composed of an alpha and beta chain, there are 6 CDR regions per TCR.

The germ line TCR β locus on chromosome 7q34 has two constant, two D, 14 J and 64 V gene segments, which are recombined during T-cell development to yield numerous VDJ recombined T-cell clones; likewise, TCR α locus on chromosome 14q11 has one constant, 61 J and 44 V segments. Further variability and antigen recognition capacity is introduced by nucleotide insertion (NI) in the recombined TCR α and β VDJ sequences. This generates a vast T-cell repertoire, yielding in excess of a trillion potential TCRαβ combinations capable of reacting to non-self (and self) peptides (PMID: 26763333)

Thus, the TCR-alpha and TCR-beta gene products together form a unique protein dimer that ultimately expressed on the cell surface as the TCR. At this point, the progenitor T or B cell is now considered a naive T or B cell, and migrates to the lymph nodes. Each TCR has the capacity to bind to a specific (or range of) antigen(s). Thus, the TCR repertoire, or all unique TCR sequences that exist within an individual at any given time, represents the total capacity of the immune system to recognize self from non-self.  

Antigens for TCR recognition are found on the surfaces of all nucleated cells (via MHC class I) or on specialized antigen-presenting cells (i.e dendritic cells, mononuclear phagocytes) (via MHC clas II). MHC-I antigens originate from peptides found within the cytoplasm of normal cells (so help clear infected cells or tumor cells), and interact with CD8+ cytotoxic T-cells. MHC-II antigens originate from peptides found within extracellular spaces (so help to clear bacterial toxins, infectious pathogens), and interact with CD4+ helper T-cells. Once a T-cell binds a presented antigen/MHC complex, it becomes activated and is stimulated to proliferate.


# **ImmunoSEQ (Adaptive biotechnologies)**

* Goal: Identifying the TCR repertoire (specifically, the TCR-beta chain CDR3 variable region) in a sample.

* Method: This is accomplished by multiplex PCR using the constant regions of the TCR sequence, followed by subsequent sequencing of libraries to identify unique nucleic acid sequences that act as 'barcodes' for each T-cell clonotype. The PCR amplification strategy has 2 steps - First, multiplex PCR using primers spanning all (?) unique VJ segments. Second, PCR adds barcodes and adapter sequences compatble with Illumina MiSeq/HiSeq platforms. PCR bias has been studied in detail, and primer concentrations adjusted so that the final PCR primer mix results in equal amplification of all possible products prior to sequencing.


* Data and analyses tools available through the immunoSEQ ANALYZER: https://clients.adaptivebiotech.com/ 



# **Study Design**

* Project Summary: 

*"Overcoming resistance to PD-1/PD-L1 blockade in non-T cell-inflamed tumors by in situ immune modulation (ISIM) with Flt3L, radiotherapy, and CD40/TLR3 agonists.*

** PD-1/PD-L1 blockade has revolutionized treatments for various malignancies, but many patients do not respond or develop resistance. Pre-existing immune state is associated with response to PD-1/PD-L1 blockade.

*** Presence of TILs
*** PD-L1 expression in the tumor microenvironment (TME)

** Flt3 (aka CD135) is an important cytokine receptor in hematopoietic differentiation. Specifically, Flt3 and its cognate ligand Flt3 are critical for the development of lymphoid cells, dendritic cells (DCs) and stem/progenitor cells.

*** Flt3L promotes the development of CD103+ DCs, which have been shown to uniquely transport intact antigens to the tumor draining lymphe node (TdLN) and prime CD8+ T cells. Expansion of this population enhances tumor responses to PD-L1 inhibition (Salmon H et al. Immunity, 2016)

** Radiotherapy activates DCs to promote presentation of tumor antigens to CD8+ T cells

** Anti-CD40 mAb and TLR3 agonist synergistically activate antigen-presenting cells to expand tumor-specific CD8_ T cell responses



* Hypothesis: 
** In situ Immunomodulation (ISIM) with Flt2L, radiotherapy and CD40/TLR3 agonists will convert non-T cell inflamed (cold) to inflamed (hot) tumors and render the tumors sensitive to PD-1/PD-L1 blockade.
** ISIM increases TCR clonality, and cyclic ISIM induces distinct TCR clones to the tumor.
  



```{r LymphoSeq: Import Data / Extract sequences}

## Load LymphoSeq package
library(LymphoSeq)
library(LymphoSeqDB)

##  IMPORTING DATA ##
# Use readImmunoSeq() function to import .tsv files from ImmunoSEQ output
Sample.list <-  readImmunoSeq(path = paste0(mount, wd, "/Data/samplesMerged"))
# Adjust names
names(Sample.list) <- gsub(".tsv", "", list.files(paste0(mount, wd, "/Data/samplesMerged")))
names(Sample.list) <- gsub("-", ".", names(Sample.list))

# Filter for clones for coverage threshold
# coverage = 5
# Sample.list <- lapply(Sample.list, subset, as.numeric(count) >= coverage)



##  EXTRACTING PRODUCTIVE SEQUENCES ##
# Aggregate productive counts by nucleotide
Sample.nt.list <- productiveSeq(file.list = Sample.list, aggregate = "nucleotide", prevalence = FALSE)
# Filter for clones for coverage threshold
coverage = 5
Sample.nt.list_filter <- lapply(Sample.nt.list, subset, as.numeric(count) >= coverage)


# Aggregate productve counts by amino acid
Sample.aa.list <- productiveSeq(file.list = Sample.list, aggregate = "aminoAcid", prevalence = TRUE)
# Filter for clones for coverage threshold
coverage = 5
Sample.aa.list_filter <- lapply(Sample.aa.list, subset, as.numeric(count) >= coverage)



## GENERATING TABLE OF AA COUNTS ##
source(paste0(mount, wd, "/Ito_AAsequences.R"))

```





```{r LymphoSeq: summary stats}

## Generate table of summary statistics (i.e. number of sequences, number of productive sequences, entropy, clonality, Gini coefficient, frequency of top productive sequences)
# For all sequences
Sample.summary <- clonality(file.list = Sample.list)
Sample.summary <- Sample.summary[order(Sample.summary$samples) ,]
# Add sample/time variables
Sample.summary$SampleID <- factor(rep(c("I_1-1", "I_1-2", "I_1-3", "I_1F-1", "I_1F-2", "I_2-1", "I_2-2", "I_2-3", "NT-1", "NT-2", "NT-3"), each = 3), levels = c("NT-1", "NT-2", "NT-3", "I_1-1", "I_1-2", "I_1-3",  "I_2-1", "I_2-2", "I_2-3", "I_1F-1", "I_1F-2"))
#
Sample.summary$TimePoint <- rep(c("1", "2", "3"), times = 11)
#
library(stringr)
Sample.summary$Group <- str_extract(Sample.summary$SampleID, pattern = "^.*-")
Sample.summary$Group <- gsub("-", "", Sample.summary$Group)
Sample.summary$Group <- factor(Sample.summary$Group, levels = c("NT", "I_1", "I_2", "I_1F"))
#
Sample.summary$Replicate <- unlist(lapply(strsplit(as.character(Sample.summary$SampleID), "-", fixed = TRUE), "[", 2))
# Re-name headers:
colnames(Sample.summary)[1:4] = c("Sample", "Unique.Rearrangements", "Productive.Rearrangements", "Total.Templates")
# Save to file
write.csv(Sample.summary, paste0(mount, wd, "/Results/Sample.summary.csv"))

# For productive sequences
Sample.aa.summary <- clonality(file.list = Sample.aa.list)
# Sort by sample id
Sample.aa.summary <- Sample.aa.summary[order(Sample.aa.summary$samples) ,]
# Add sample/time variables
Sample.aa.summary$SampleID <- factor(rep(c("I_1-1", "I_1-2", "I_1-3", "I_1F-1", "I_1F-2", "I_2-1", "I_2-2", "I_2-3", "NT-1", "NT-2", "NT-3"), each = 3), levels = c("NT-1", "NT-2", "NT-3", "I_1-1", "I_1-2", "I_1-3",  "I_2-1", "I_2-2", "I_2-3", "I_1F-1", "I_1F-2"))
Sample.aa.summary$TimePoint <- rep(c("1", "2", "3"), times = 11)


## Generate Overall Metrics Plots ##
Sample.summary.melt <- melt(Sample.summary)
#
for(i in unique(Sample.summary.melt$variable)){
# Plot clonality
ggplot(data = subset(Sample.summary.melt, variable == i), aes(x = TimePoint, y = value, fill = Group, color = Group)) +
   geom_line(aes(group = Replicate), color = "black", alpha = 0.1) +
   geom_point() + 
   geom_boxplot(alpha = 0.2) +
   ylab(i) +
   facet_grid(~Group) + 
   theme(text=element_text(size = 20), legend.text=element_text(size = 20))
ggsave(filename = paste0(mount, wd, "/Figures/OverallMetrics_", i, ".jpeg"), dpi = 300, units = "in", height = 4, width = 10)
}


```


```{r repertoire similarity / diversity}

## Calculate 'Clonal Relatedness'
Sample.ClonalRelatedness <- clonalRelatedness(list = Sample.list, editDistance = 10)
Sample.productive.ClonalRelatedness <- clonalRelatedness(list = Sample.nt.list, editDistance = 10)


## Perform Multiple Sequence Alignment
alignSeq(list = Sample.aa.list, sample = "Tumor", type = "aminoAcid", 
         method = "ClustalW", output = "console")

## Search for a sequence across samples using searchSeq() function
# Can use the editDistance argument to search for similar sequences
searchSeq(list = Sample.aa.list, sequence = "CASTPNYSNQPQHF", type = "aminoAcid", 
          match = "global", editDistance = 0)


## Search for sequences with known antigen specificity
published <- searchPublished(list = Sample.aa.list)





## VISUALIZING REPERTOIRE DIVERSITY ##

## Lorenz curve

# For a single sample:
lorenzCurve(samples = names(Sample.aa.list)[31:33], list = Sample.aa.list[31:33])

# For all samples:
lorenzCurve(samples = names(Sample.aa.list), list = Sample.aa.list)


## Top sequences
# Display the number of sequences at the top of bar plot and add a title
n <- as.character(lapply(Sample.aa.list_filter, nrow))
topSeqsPlot(list = Sample.aa.list_filter, top = 30) + 
   ggplot2::annotate("text", x = 1:length(Sample.aa.list_filter), y = 105, label = n, color = "black") +
   ggplot2::expand_limits(y = c(0, 110)) + ggplot2::ggtitle("Top 10 Clones") + 
   ggplot2::scale_x_discrete(limits = names(Sample.aa.list_filter)) +
   ggplot2::theme(axis.title=element_text(size = 25), axis.text.x=element_text(size = 25), axis.text.y=element_text(size = 15))






## Sample Similarity Measures ##

# To compare the T or B cell repertoires of all samples in a pairwise fashion, use the bhattacharyyaMatrix or  similarityMatrix functions. Both the Bhattacharyya coefficient and similarity score are measures of the amount of overlap between two samples. The value for each ranges from 0 to 1 where 1 indicates the sequence frequencies are identical in the two samples and 0 indicates no shared frequencies. The Bhattacharyya coefficient differs from the similarity score in that it involves weighting each shared sequence in the two distributions by the arithmetic mean of the frequency of each sequence, while calculating the similarity scores involves weighting each shared sequence in the two distributions by the geometric mean of the frequency of each sequence in the two distributions.





## MORISITA INDEX
# Using pairwisePlot() function for single mouse
# Using count data frame generated above
# sample
sample <- "NT.3"
MH <- vegdist(t(TotFreqDF.wide[, grep(sample, colnames(TotFreqDF.wide))]), method = "horn")
# Convert to matrix
MH <- as.data.frame(as.matrix(MH))
# Plot
pairwisePlot(matrix = 1-MH) + 
   ggplot2::scale_fill_gradientn(limits = c(0,1), colors = c("#deebf7", "#4eee94", "#ffc125", "#cd0000")) + 
   ggplot2::labs(fill = "Similarity score") #+
   #ggplot2::geom_text(aes(label = round(Score, digits = 2)))



# Using count data frame generated above
MH <- vegdist(t(TotFreqDF.wide), method = "horn")
# Convert to matrix
MH <- as.data.frame(as.matrix(MH))
# Add color annotations
Group <- ifelse(grepl("^NT", colnames(MH)), "NT", "I")
Sample <- ifelse(grepl("\\.1$", colnames(MH)), "1",
                 ifelse(grepl("\\.2$", colnames(MH)), "2", "3"))
ann.cols <- data.frame("Group" = Group, "Sample" = Sample)
rownames(ann.cols) = colnames(MH)
# Set annotation colors
ann.colors <- list("Group" = c("I" = "blue", "NT" = "grey30"),
                   "Sample" = c("1" = "#FFB6C1", "2" = "red", "3" = "darkred"))
# Heatmap of MH similarity index
require(pheatmap)
require(RColorBrewer)
colors <- colorRampPalette(rev(brewer.pal(9, "Reds")) )(255)
labels <- labels
# Plot
pheatmap(MH,
         annotation_names_row = FALSE,
         annotation_names_col = FALSE,
         annotation_colors = ann.colors,
         #annotation_row = ann_rows,
         annotation_col = ann.cols,
         clustering_distance_cols = "correlation", clustering_method = "complete",
         show_rownames = TRUE,
         show_colnames = TRUE,
         border_color = "NA",
         col=colors)


```





```{r LymphoSeq: Differential Abundance}


## DETERMINE DIFFERENTIAL ABUNDANCE ##

NT.1.1_NT.1.2 <- differentialAbundance(list = Sample.aa.list, 
                      sample1 = "NT.1.1", 
                      sample2 = "NT.1.2", 
                      type = "aminoAcid", 
                      q = 1,
                      zero = 0.001)

NT.1.2_NT.1.3 <- differentialAbundance(list = Sample.aa.list, 
                      sample1 = "NT.1.2", 
                      sample2 = "NT.1.3", 
                      type = "aminoAcid", 
                      q = 1,
                      zero = 0.001)


NT.2.1_NT.2.2 <- differentialAbundance(list = Sample.aa.list, 
                      sample1 = "NT.2.1", 
                      sample2 = "NT.2.2", 
                      type = "aminoAcid", 
                      q = 1,
                      zero = 0.001)

NT.2.2_NT.2.3 <- differentialAbundance(list = Sample.aa.list, 
                      sample1 = "NT.2.2", 
                      sample2 = "NT.2.3", 
                      type = "aminoAcid", 
                      q = 1,
                      zero = 0.001)

NT.3.1_NT.3.2 <- differentialAbundance(list = Sample.aa.list, 
                      sample1 = "NT.3.1", 
                      sample2 = "NT.3.2", 
                      type = "aminoAcid", 
                      q = 1,
                      zero = 0.001)

NT.3.2_NT.3.3 <- differentialAbundance(list = Sample.aa.list, 
                      sample1 = "NT.3.2", 
                      sample2 = "NT.3.3", 
                      type = "aminoAcid", 
                      q = 1,
                      zero = 0.001)

I_1.1.1_I_1.1.2 <- differentialAbundance(list = Sample.aa.list, 
                      sample1 = "I_1.1.1", 
                      sample2 = "I_1.1.2", 
                      type = "aminoAcid", 
                      q = 1,
                      zero = 0.001)

I_1.1.2_I_1.1.3 <- differentialAbundance(list = Sample.aa.list, 
                      sample1 = "I_1.1.2", 
                      sample2 = "I_1.1.3", 
                      type = "aminoAcid", 
                      q = 1,
                      zero = 0.001)

I_1.2.1_I_1.2.2 <- differentialAbundance(list = Sample.aa.list, 
                      sample1 = "I_1.2.1", 
                      sample2 = "I_1.2.2", 
                      type = "aminoAcid", 
                      q = 1,
                      zero = 0.001)

I_1.2.2_I_1.2.3 <- differentialAbundance(list = Sample.aa.list, 
                      sample1 = "I_1.2.2", 
                      sample2 = "I_1.2.3", 
                      type = "aminoAcid", 
                      q = 1,
                      zero = 0.001)

I_1.3.1_I_1.3.2 <- differentialAbundance(list = Sample.aa.list, 
                      sample1 = "I_1.3.1", 
                      sample2 = "I_1.3.2", 
                      type = "aminoAcid", 
                      q = 1,
                      zero = 0.001)

I_1.3.2_I_1.3.3 <- differentialAbundance(list = Sample.aa.list, 
                      sample1 = "I_1.3.2", 
                      sample2 = "I_1.3.3", 
                      type = "aminoAcid", 
                      q = 1,
                      zero = 0.001)

I_2.1.1_I_2.1.2 <- differentialAbundance(list = Sample.aa.list, 
                      sample1 = "I_2.1.1", 
                      sample2 = "I_2.1.2", 
                      type = "aminoAcid", 
                      q = 1,
                      zero = 0.001)

I_2.1.2_I_2.1.3 <- differentialAbundance(list = Sample.aa.list, 
                      sample1 = "I_2.1.2", 
                      sample2 = "I_2.1.3", 
                      type = "aminoAcid", 
                      q = 1,
                      zero = 0.001)

I_2.2.1_I_2.2.2 <- differentialAbundance(list = Sample.aa.list, 
                      sample1 = "I_2.2.1", 
                      sample2 = "I_2.2.2", 
                      type = "aminoAcid", 
                      q = 1,
                      zero = 0.001)

I_2.2.2_I_2.2.3 <- differentialAbundance(list = Sample.aa.list, 
                      sample1 = "I_2.2.2", 
                      sample2 = "I_2.2.3", 
                      type = "aminoAcid", 
                      q = 1,
                      zero = 0.001)

I_2.3.1_I_2.3.2 <- differentialAbundance(list = Sample.aa.list, 
                      sample1 = "I_2.3.1", 
                      sample2 = "I_2.3.2", 
                      type = "aminoAcid", 
                      q = 1,
                      zero = 0.001)

I_2.3.2_I_2.3.3 <- differentialAbundance(list = Sample.aa.list, 
                      sample1 = "I_2.3.2", 
                      sample2 = "I_2.3.3", 
                      type = "aminoAcid", 
                      q = 1,
                      zero = 0.001)

I_1F.1.1_I_1F.1.2 <- differentialAbundance(list = Sample.aa.list, 
                      sample1 = "I_1F.1.1", 
                      sample2 = "I_1F.1.2", 
                      type = "aminoAcid", 
                      q = 1,
                      zero = 0.001)

I_1F.1.2_I_1F.1.3 <- differentialAbundance(list = Sample.aa.list, 
                      sample1 = "I_1F.1.2", 
                      sample2 = "I_1F.1.3", 
                      type = "aminoAcid", 
                      q = 1,
                      zero = 0.001)

I_1F.2.1_I_1F.2.2 <- differentialAbundance(list = Sample.aa.list, 
                      sample1 = "I_1F.2.1", 
                      sample2 = "I_1F.2.2", 
                      type = "aminoAcid", 
                      q = 1,
                      zero = 0.001)

I_1F.2.2_I_1F.2.3 <- differentialAbundance(list = Sample.aa.list, 
                      sample1 = "I_1F.2.2", 
                      sample2 = "I_1F.2.3", 
                      type = "aminoAcid", 
                      q = 1,
                      zero = 0.001)

```


